nexus/deployment/src/blueprint_builder.rs

```diff
@@ -0,0 +1,682 @@
+                        generation: Generation::new(),
+                        zones: vec![],
+                    });
+                (*sled_id, zones)
+            })
+            .collect();
+        Blueprint {
+            id: Uuid::new_v4(),
+            omicron_zones: omicron_zones,
+            zones_in_service: self.zones_in_service,
+            parent_blueprint_id: Some(self.parent_blueprint.id),
+            time_created: chrono::Utc::now(),
+            creator: self.creator,
+            comment: self.comments.join(", "),
+        }
+    }
+
+    /// Sets the blueprints "comment"
+    ///
+    /// This is a short human-readable string summarizing the changes reflected
+    /// in the blueprint.  This is only intended for debugging.
+    pub fn comment(&mut self, comment: &str) {
+        self.comments.push(comment.to_owned());
+    }
+
+    pub fn sled_ensure_zone_internal_ntp(
+        &mut self,
+        sled_id: Uuid,
+    ) -> Result<bool, Error> {
+        // If there's already an NTP zone on this sled, do nothing.
+        let has_ntp = self
+            .parent_blueprint
+            .omicron_zones
+            .get(&sled_id)
+            .map(|found_zones| {
+                found_zones.zones.iter().any(|z| {
+                    matches!(
+                        z.zone_type,
+                        OmicronZoneType::BoundaryNtp { .. }
+                            | OmicronZoneType::InternalNtp { .. }
+                    )
+                })
+            })
+            .unwrap_or(false);
+        if has_ntp {
+            return Ok(false);
+        }
+
+        let sled_info = self.sled_resources(sled_id)?;
+        let sled_subnet = sled_info.subnet;
+        let ip = self.sled_alloc_ip(sled_id)?;
+        let ntp_address = SocketAddrV6::new(ip, NTP_PORT, 0, 0);
+
+        // Construct the list of internal DNS servers.
+        //
+        // It'd be tempting to get this list from the other internal NTP
+        // servers but there may not be any of those.  We could also
+        // construct this list manually from the set of internal DNS servers
+        // actually deployed.  Instead, we take the same approach as RSS:
+        // these are at known, fixed addresses relative to the AZ subnet
+        // (which itself is a known-prefix parent subnet of the sled subnet).
+        let dns_servers =
+            get_internal_dns_server_addresses(sled_subnet.net().network());
+
+        // The list of boundary NTP servers is not necessarily stored
+        // anywhere (unless there happens to be another internal NTP zone
+        // lying around).  Recompute it based on what boundary servers
+        // currently exist.
+        let ntp_servers = self
+            .parent_blueprint
+            .all_omicron_zones()
+            .filter_map(|(_, z)| {
+                if matches!(z.zone_type, OmicronZoneType::BoundaryNtp { .. }) {
+                    Some(format!("{}.host.{}", z.id, DNS_ZONE))
```

Comment:
Fixed in c58c7edae525570e703cb113075211ccd9903c6c.  There was already some skeleton for this in the `internal-dns`  #crate.  I did have to rework some of this and since they're often used as strings it's not as type-checked as would be nice.  Extra review on this would be helpful (although I expect the end-to-end test job would fail if I somehow got a large class of DNS names wrong).

Reply:
I'm not super familiar with these, but the changes you made look good, and I don't see any lingering, now-suspect `DNS_ZONE` references. :+1: